lib/SIL/SILVTable.cpp

```diff
@@ -33,27 +33,30 @@ SILVTable *SILVTable::create(SILModule &M, ClassDecl *Class,
   SILVTable *vt = ::new (buf) SILVTable(Class, Entries);
   M.vtables.push_back(vt);
   M.VTableMap[Class] = vt;
+  // Update the Module's cache with new vtable + vtable entries:
+  for (auto &entry : Entries) {
+    M.VTableEntryCache.insert({{vt, entry.first}, entry.second});
+  }
   return vt;
 }
 
 SILFunction *
 SILVTable::getImplementation(SILModule &M, SILDeclRef method) const {
-  // FIXME: We should build a sidetable cache in the module. Linear lookup here
-  // is lame.
-  
-  for (auto &entry : getEntries()) {
-    // Check whether this mapping matches either the given decl directly or
-    // one of its overridden decl.
-    SILDeclRef m = method;
-    do {
-      if (entry.first == m)
-        return entry.second;
-    } while ((m = m.getOverridden()));
-  }
-  
+  SILDeclRef m = method;
+  do {
+    auto entry = M.VTableEntryCache.find(std::make_pair(this, m));
+    if (entry != M.VTableEntryCache.end()) {
+      return (*entry).second;
+    }
+  } while ((m = m.getOverridden()));
   return nullptr;
 }
 
+void SILVTable::removeFromVTableCache(Pair &entry) {
+  SILModule &M = entry.second->getModule();
+  M.VTableEntryCache.erase(std::make_pair(this, entry.first));
```

Comment:
If you want, you can also remove the make_pair here and just use {this, entry.first}.